x/nexus/keeper/rate_limit.go

```diff
@@ -0,0 +1,140 @@
+package keeper
+
+import (
+	"fmt"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/axelarnetwork/axelar-core/utils"
+	"github.com/axelarnetwork/axelar-core/utils/events"
+	"github.com/axelarnetwork/axelar-core/utils/key"
+	"github.com/axelarnetwork/axelar-core/x/nexus/exported"
+	"github.com/axelarnetwork/axelar-core/x/nexus/types"
+	"github.com/axelarnetwork/utils/funcs"
+)
+
+// RateLimitTransfer applies a rate limit to transfers, and returns an error if the rate limit is exceeded
+func (k Keeper) RateLimitTransfer(ctx sdk.Context, chain exported.ChainName, asset sdk.Coin, direction exported.TransferDirection) error {
+	rateLimit, found := k.getRateLimit(ctx, chain, asset.Denom)
+	// If rate limit is not set, it is treated as unbounded
+	if !found {
+		return nil
+	}
+
+	epoch := uint64(ctx.BlockTime().UnixNano() / rateLimit.Window.Nanoseconds())
+
+	transferEpoch, found := k.getTransferEpoch(ctx, chain, asset.Denom, direction)
+	// use a new transfer epoch if there was none or if the epoch is outdated
+	if !found || transferEpoch.Epoch != epoch {
+		transferEpoch = types.TransferEpoch{
+			Chain:     chain,
+			Amount:    sdk.NewCoin(asset.Denom, sdk.ZeroInt()),
+			Epoch:     epoch,
+			Direction: direction,
+		}
+	}
+
+	transferEpoch.Amount = transferEpoch.Amount.Add(asset)
+
+	if transferEpoch.Amount.Amount.GT(rateLimit.Limit.Amount) {
+		err := fmt.Errorf("transfer %s for chain %s (%s) exceeded rate limit %s with transfer rate %s", asset, transferEpoch.Chain, transferEpoch.Direction, rateLimit.Limit, transferEpoch.Amount)
+		k.Logger(ctx).Error(err.Error(), types.AttributeKeyChain, transferEpoch.Chain, types.AttributeKeyAsset, asset, types.AttributeKeyLimit, rateLimit.Limit, types.AttributeKeyTransferEpoch, transferEpoch.Amount)
```

Comment:
break this into multiple lines. It's currently way too hard to see the key/value pairs.